staging/src/k8s.io/apiserver/pkg/endpoints/filters/storageversion.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog"
+)
+
+type Barrier interface {
+	SetComplete()
+	Completed() bool
+}
+
+var _ Barrier = &neverCompleteBarrier{}
+
+type neverCompleteBarrier struct{}
+
+func (_ *neverCompleteBarrier) SetComplete()    {}
+func (_ *neverCompleteBarrier) Completed() bool { return false }
+
+var _ Barrier = &StorageVersionBarrier{}
+
+// StorageVersionBarrier indicates if the aggregator, kube-apiserver, and the
+// apiextensions-apiserver have completed reporting their storage versions.
+type StorageVersionBarrier struct {
+	mu        sync.RWMutex
+	completed bool
+}
+
+// SetComplete sets completed to true.
+func (svb *StorageVersionBarrier) SetComplete() {
+	svb.mu.Lock()
+	defer svb.mu.Unlock()
+	svb.completed = true
+}
+
+// Completed returns completed.
+func (svb *StorageVersionBarrier) Completed() bool {
+	svb.mu.RLock()
+	defer svb.mu.RUnlock()
+	return svb.completed
+}
+
+// WithStorageVersionPrecondition checks if the storage version barrier has
+// completed, if not, it only passes the following API requests:
+// 1. read requests,
+// 2. TODO: non-persisting write requests,
+// 3. kube-apiserver's write requests to the storageversion API,
+func WithStorageVersionPrecondition(handler http.Handler, svb Barrier) http.Handler {
+	if svb == nil {
+		klog.Warningf("Storage Version barrier is disabled")
+		return handler
+	}
+	var completed bool
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		fmt.Println("CHAO: in filter")
+		if completed {
```

Comment:
pretty sure that the `Handle` method is called on multiple threads, making this variable access a race